Adding IAM permissions management use-case (#912)
port-labs/port-docs: docs/build-your-software-catalog/sync-data-to-catalog/cloud-providers/aws/iam-permissions-guide.md

---
sidebar_position: 5
---

> # AWS IAM permission management

Reviewer: ```suggestion
# AWS IAM Permission Management
```

Developers need access to different cloud resources on a daily basis. With all of your most important AWS resources being exported to Port, you can create an experience for requesting IAM permissions for your developers.

It is important to be able to keep track of the permissions being allocated to your developers, whether it is who requested the permissions, or what permissions they requested.

In this step-by-step guide, you wou will create Port blueprints and actions, which will allow you to request and revoke IAM permissions for different AWS resources using Port. You will also be able to keep track of which permissions were requested, and who requested them.

## Prerequisites
- In your Github organization, create a new repository called `port-iam-permissions`. This repository will be used to hold our Github workflows.
- Install Port's GitHub app by clicking [here](https://github.com/apps/getport-io/installations/new). Make sure to give the Port Github app permissions for the `port-iam-permissions` repository.

> - In your AWS console, [create an IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html) called `port-iam-management-user` with the following IAM permission policy:

Reviewer: ```suggestion
- In your AWS console, [create an IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html) called `port-iam-management-user` with the following IAM permissions policy:
```

---
sidebar_position: 5
tags:
  - AWS
  - IAM Permissions
  - Guide
  - Github Actions
---

# AWS IAM Permission Management

## Overview

Developers need access to different cloud resources on a daily basis. With all of your most important AWS resources being exported to Port, you can create an experience for requesting IAM permissions for your developers directly from Port.

It is important to be able to keep track of the permissions being allocated to your developers, whether it is who requested the permissions, or what permissions they requested.

In this step-by-step guide, you will create Port blueprints and actions, which will allow you to request and revoke IAM permissions for different AWS resources using Port. You will also be able to keep track of which permissions were requested, and who requested them.

## Prerequisites

> - Prepare your Port organization's `Client ID` and `Client Secret`. To find you Port credentials, click [here](/docs/build-your-software-catalog/sync-data-to-catalog/api/api.md#find-your-port-credentials)!

Reviewer 2: ```suggestion
- Prepare your Port organization's `Client ID` and `Client Secret`. To find you Port credentials, click [here](/docs/build-your-software-catalog/sync-data-to-catalog/api/api.md#find-your-port-credentials).
```

                    "title": "Sign-in URL",
                    "type": "string",
                    "description": "The sign-in URL for this temporary permission",
                    "format": "url"
                },
                "policy_arn": {
                    "title": "Policy ARN",
                    "type": "string",
                    "icon": "DefaultProperty"
                },
                "role_arn": {
                    "title": "Role ARN",
                    "type": "string",
                    "icon": "DefaultProperty"
                }
            },
            "required": []
        },
        "mirrorProperties": {},
        "calculationProperties": {},
        "aggregationProperties": {},
        "relations": {
            "permissions": {
                "title": "Permissions",
                "target": "iam_permissions",
                "required": false,
                "many": true
            },
            "aws_resource": {
                "title": "AWS Resource",
                "target": "aws_resource",
                "required": false,
                "many": false
            }
        }
    }
    ```
    </details>

> :::note

Reviewer 2: plz add a title to the admonition.
Consider using a type other than `note`, it is quite easy to miss.

Author: I think this is a note, if this gets missed it shouldn't affect the use-case. 
I the `Summary` section, I reference this again.

This isn't supposed to catch too much attention

For this guide's simplicity, the blueprints above have pre-defined options for resource types, which are `EC2` and `S3`. 

The blueprints can be modified to support for any type of AWS resource by adding extra options to the `resource_type` properties, both in the `AWS Resource` and the `IAM Permissions` blueprints.
:::

<p align="center">
<img src='/img/build-your-software-catalog/sync-data-to-catalog/cloud-providers/aws/iam-permissions-data-model.png' width='75%' border='1px' />
</p>

## Actions

> We want to be able to provision and revoke permissions for AWS resources from Port. To do so, we will need to create some [Port actions](/docs/create-self-service-experiences/create-self-service-experiences.md) in our Port organization, and set up some action backends.

Reviewer 2: ```suggestion
We want to be able to provision and revoke permissions for AWS resources from Port. To do so, we will need to create some [actions](/docs/create-self-service-experiences/create-self-service-experiences.md) in our Port organization, and set up some action backends.
```

                with:
                    clientId: ${{ secrets.PORT_CLIENT_ID }}
                    clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
                    identifier: ${{ env.POLICY_NAME }}
                    title: ${{ env.POLICY_NAME }}
                    blueprint: provisioned_permissions
                    properties: |
                    {
                        "iam_policy": ${{ steps.create-variables.outputs.POLICY }},
                        "requester": "${{ fromJson(inputs.port_payload).trigger.by.user.email }}",
                        "sign_in_url": "${{ steps.create-variables.outputs.SIGN_IN_URL }}",
                        "role_arn": ${{ steps.apply-policies.outputs.ROLE_ARN }},
                        "policy_arn": ${{ steps.apply-policies.outputs.POLICY_ARN }}
                    }
                    relations: |
                    {
                        "aws_resource": "${{ fromJson(inputs.port_payload).context.entity }}",
                        "permissions": ${{ steps.create-jsons.outputs.PERMISSIONS_ARRAY }}
                    }
                - uses: port-labs/port-github-action@v1

>                 with:

Author: Since this is Github actions, it will end automatically thanks to the integration. It's another block of code that is not really needed

    {
        "identifier": "revoke_permissions",
        "title": "Revoke permissions",
        "icon": "Alert",
        "userInputs": {
            "properties": {},
            "required": []
        },
        "invocationMethod": {
            "type": "GITHUB",
            "org": "<YOUR_GITHUB_ORG>",
            "repo": "port-iam-permissions",
            "workflow": "delete-iam-permissions.yaml",
            "omitUserInputs": true,
            "omitPayload": false,
            "reportWorkflowStatus": true
        },
        "trigger": "DELETE",
        "description": "Revokes IAM permissions",
        "requiredApproval": false
    }
    ```
</details>

## Manage permissions using Port
Before we get to provisioning and revoking permissions, we have 2 things to complete:
1. Define which AWS resources we want provision permissions for.
2. Define which permissions we want to allow our developers to request and provision.

### Defining AWS resources
Managing the AWS resources we want to provision permissions for will be done via Port entities. Navigate to the [AWS Resources](https://app.getport.io/aws_resources) catalog page to create some example entities.

> :::note

Reviewer 2: Title

Let's create 2 `IAM Permissions` entities:

1. We will create one entity with the resource type `S3`:
    * Title: `Put S3 objects`
    * Identifier: `s3:PutObject`
    * Resource Type: `S3`

2. We will create another entity with the resource type `EC2`:
    * Title: `Stop EC2 Instance`
    * Identifier: `ec2:StopInstances`
    * Resource Type: `EC2`

:::note
Feel free to add more IAM permissions of your own. Make sure that the identifier of the entity matches the IAM permission you want to add.
:::
 
<p align="center">
<img src='/img/build-your-software-catalog/sync-data-to-catalog/cloud-providers/aws/iam-permissions-create-iam-permissions-entity.png' width='50%' border='1px' />
</p>

We are all set!

### Managing permissions
Now that we finished setting up our Port environment, and our action backends, we are ready to manage IAM permissions for our AWS resources!

#### Provision permissions
Let's start by creating new temporary permissions for our S3 bucket `my-s3-bucket`, using Port. Navigate to the bucket's [entity page](https://app.getport.io/aws_resourceEntity?identifier=arn%3Aaws%3As3%3A%3A%3Amy-s3-bucket).

> Click on the `...` at the top right of the entity screen -> click `Request permissions` -> choose the `s3:PutObject` -> click `Execute`.

Reviewer 3: ```suggestion
Click on the `...` at the top right of the entity screen -> click `Request permissions` -> choose the `s3:PutObject` permission -> click `Execute`.
```

This will trigger a new action run which will appear in the right action runs bar. Click on the action run to navigate to the run page.

> When the Port action run will end, you will get actions logs showing you info regarding:

Reviewer 3: ```suggestion
When the Port action run will end, you will get action logs which will show you info regarding:
```

- Which AWS resource were the IAM permissions provisioned for.
- Who requested the IAM permissions.
- The sign-in URL for the provisioned role.

The actoin will also create a new `Provisioned Permissions` entity which you can see in the [Provisioned Permissions](https://app.getport.io/provisioned_permissions) catalog page.

> Copy the sign-in URL and paste it tou your browser URL bar. Click the `Switch Role` button. 

Reviewer 3: ```suggestion
Copy the sign-in URL and paste it to your browser's URL bar. Click the `Switch Role` button. 
```

> You are now signed in to your new role, which has permissions as defined the the Port action! 🥳

Reviewer 3: ```suggestion
You are now signed in to your new role, which has permissions as defined in the Port action! 🥳
```

<p align="center">

> <img src='/img/build-your-software-catalog/sync-data-to-catalog/cloud-providers/aws/iam-permissions-provision-permissions-action.png' width='75%' border='1px' />

Reviewer 3: Can we add an image where the email address is something that maybe ends with getport.io?

</p>
#### Revoke permissions
Now, we want to revoke the permissions we provisioned for our S3 bucket `my-s3-bucket`. Let's navigate to our new `Provisioned Permission` entity.

Navigate to the [Provisioned Permissions](https://app.getport.io/provisioned_permissions) catalog page -> Click on the new `Permission-XXXXXXXX` entity.

> :::note

Reviewer 3: change to tip and add title - "about provisioned permissions"

In the `Provisioned Permissions` entity page, you can learn important infromation regarding the provisioned permissions. For example, you can see the generated IAM policy and the permissions requester.

In the `Related entities` section, you can see which permissions were provisioned, and to which resource.
:::

Click on the `...` at the top right of the entity screen -> click `Revoke permissions` -> click `Delete`.

This will trigger a new action run which will appear in the right action runs bar. Click on the action run to navigate to the run page.

> When the Port action run will end, you will get actions logs showing you info regarding:

Reviewer 3: ```suggestion
When the Port action run will end, you will get action logs which will show you info regarding:
```

Feel free to further experiment with the use-case by adding more `IAM Permissions` and `AWS Resources` entities. You can also add more resource types by modifying the `resource_type` property in the `IAM Permissions` and `AWS Resources` blueprints.

See the [Next Steps](#next-steps) section to understand how to take this guide one step further with your Port environment.

## Next Steps
- **Install Port's [AWS exporter](/docs/build-your-software-catalog/sync-data-to-catalog/cloud-providers/aws/aws.md)** - You can use Port's AWS exporter to automatically populates your software catalog from your AWS environement. You can use the AWS exporter to populate your `AWS Resources` blueprints with different AWS resources.
- **Enforce [manual approval](/docs/create-self-service-experiences/set-self-service-actions-rbac/set-self-service-actions-rbac.md#configure-manual-approval-for-actions--give-approval-permissions) for your Port actions** - To have control over who is provisioning and revoking permissions, you can set up manual approval for your actions. This will enable you to provide a request-approve flow for provisioning and revoking permissions using Port.

> - ***Coming soon*⏱️: Temporary permissions using [Automations](https://roadmap.getport.io/ideas/p/automation) and the `Timer` property** - With the automations feature, you will be able to automatically trigger actions using events from the catalog. You can use the `Timer Expired` event to trigger the `Revoke permissions` action and create a temporary permissions experience.

Reviewer 2: This is cool, but it's very easy to forget to change it, plz make sure you remember to come back here when the feature is ready

> - ***Coming soon*⏱️: Temporary permissions using [Automations](https://roadmap.getport.io/ideas/p/automation) and the `Timer` property** - With the automations feature, you will be able to automatically trigger actions using events from the catalog. You can use the `Timer Expired` event to trigger the `Revoke permissions` action and create a temporary permissions experience.

Reviewer 3: ```suggestion
- ***Coming soon* ⏱️: Temporary permissions using [Automations](https://roadmap.getport.io/ideas/p/automation) and the `Timer` property** - With the automations feature, you will be able to automatically trigger actions using events from the catalog. You can use the `Timer Expired` event to trigger the `Revoke permissions` action and create a temporary permissions experience.
```
